src/libraries/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs

```diff
@@ -9,171 +9,162 @@
-        {
-            using (Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
-            {
-                AssertExtensions.Throws<ArgumentNullException>("asyncResult", () => s.EndDisconnect(null));
-                AssertExtensions.Throws<ArgumentNullException>("e", () => s.DisconnectAsync(null));
-                AssertExtensions.Throws<ArgumentException>("asyncResult", () => s.EndDisconnect(Task.CompletedTask));
-                s.Dispose();
-                Assert.Throws<ObjectDisposedException>(() => s.Disconnect(true));
-                Assert.Throws<ObjectDisposedException>(() => s.BeginDisconnect(true, null, null));
-                Assert.Throws<ObjectDisposedException>(() => s.EndDisconnect(null));
-                Assert.Throws<ObjectDisposedException>(() => { s.DisconnectAsync(null); });
-            }
-        }
+        protected Disconnect(ITestOutputHelper output) : base(output) { }
 
         [Theory]
         [InlineData(true)]
         [InlineData(false)]
-        [OuterLoop("https://github.com/dotnet/runtime/issues/18406")]
-        public void Disconnect_Success(bool reuseSocket)
+        public async Task Disconnect_Success(bool reuseSocket)
         {
-            AutoResetEvent completed = new AutoResetEvent(false);
-
             IPEndPoint loopback = new IPEndPoint(IPAddress.Loopback, 0);
             using (var server1 = SocketTestServer.SocketTestServerFactory(SocketImplementationType.Async, loopback))
             using (var server2 = SocketTestServer.SocketTestServerFactory(SocketImplementationType.Async, loopback))
             {
-                SocketAsyncEventArgs args = new SocketAsyncEventArgs();
-                args.Completed += OnCompleted;
-                args.UserToken = completed;
-                args.RemoteEndPoint = server1.EndPoint;
-                args.DisconnectReuseSocket = reuseSocket;
-
                 using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
                 {
-                    if (client.ConnectAsync(args))
-                    {
-                        completed.WaitOne();
-                    }
-
-                    Assert.Equal(SocketError.Success, args.SocketError);
-
-                    client.Disconnect(reuseSocket);
+                    await ConnectAsync(client, server1.EndPoint);
+                    Assert.True(client.Connected);
 
+                    await DisconnectAsync(client, reuseSocket);
                     Assert.False(client.Connected);
 
-                    args.RemoteEndPoint = server2.EndPoint;
-
-                    if (client.ConnectAsync(args))
+                    if (reuseSocket)
                     {
-                        completed.WaitOne();
+                        // Note that the new connect operation must be asynchronous
+                        // (why? I'm not sure, but that's the way it works currently)
```

Comment:
If this is expected behavior, I guess it should be documented. Do we know someone who understands this better?

Reply:
I don't know. We seem to be enforcing this limitation ourselves, though I don't know if it's because DisconnectEx does not support it, or for some other (possibly legacy and no longer relevant) reason.

I skimmed the DisconnectEx docs on MSDN and didn't see anything that seemed relevant here.

I'm inclined to leave this as is currently and file a separate issue to investigate further. Sound reasonable?

Reply:
Makes sense to me, since the question is not strictly related to the PR.

Reply:
I added an issue here: https://github.com/dotnet/runtime/issues/47206